CRM/Core/BAO/CustomGroup.php

```diff
@@ -28,6 +28,80 @@ public static function self_hook_civicrm_post(\Civi\Core\Event\PostEvent $e): vo
     Civi::cache('metadata')->flush();
   }
 
+  /**
+   * Retrieve all enabled custom groups and fields in a nested array.
+   *
+   * @return array[]
+   */
+  public static function getActive(): array {
+    $allGroups = self::getAll();
+    $allGroups = array_values(array_filter($allGroups, fn($group) => $group['is_active']));
+    foreach ($allGroups as $groupIndex => $group) {
+      $allGroups[$groupIndex]['fields'] = array_values(array_filter($group['fields'], fn($field) => $field['is_active']));
+    }
+    return $allGroups;
+  }
+
+  /**
+   * Same as self::getActive() but returns only active groups and
+   * fields which the given user is allowed to see.
+   *
+   * @param int $permissionType (CRM_Core_Permission::VIEW or CRM_Core_Permission::EDIT)
+   * @param int|null $userId
+   * @return array[]
+   */
+  public static function getPermitted(int $permissionType, ?int $userId = NULL) {
+    if (!in_array($permissionType, [CRM_Core_Permission::EDIT, CRM_Core_Permission::VIEW], TRUE)) {
+      throw new CRM_Core_Exception('permissionType must be CRM_Core_Permission::VIEW or CRM_Core_Permission::EDIT');
+    }
+    $allGroups = self::getActive();
+    $allowedGroupIds = CRM_Core_Permission::customGroup($permissionType, FALSE, $userId);
+    $allGroups = array_filter($allGroups, function($group) use ($allowedGroupIds) {
+      return in_array($group['id'], $allowedGroupIds);
+    });
+    return array_values($allGroups);
+  }
+
+  /**
+   * Fetch all custom groups and fields in a nested array.
+   *
+   * Avoids calling the API to prevent recursion or early-bootstrap issues.
+   *
+   * @return array[]
+   */
+  public static function getAll(): array {
+    $cacheString = __CLASS__ . __FUNCTION__ . '_' . CRM_Core_I18n::getLocale();
+    $custom = Civi::cache('metadata')->get($cacheString);
```

Comment:
I'm not advocating changing this but just want to note that when I did performance work on WMF serializing & unserializing arrays was more expensive that running queries. We have this pattern elsewhere & the theory is that as long as we only do it once per request then all good but it could turn out in the future that someone makes the case to use `Civi::statics` in places like this instead. 

We do tend to use `if Civi::cache('metadata')->has($cacheString)) ` for the checks in similar code but there is no real difference (there was a performance difference but that got fixed)

Reply:
It would be interesting to see a benchmark between using the cache vs fetching from the db & stashing in memory.

Reply:
> @eileenmcnaughton We do tend to use if Civi::cache('metadata')->has($cacheString))  for the checks in similar code but there is no real difference (there was a performance difference but that got fixed)

Well, that function has a comment saying not to use it:
https://github.com/civicrm/civicrm-core/blob/f452d72c08c27b1881af42f12c669185c43c3e30/CRM/Utils/Cache/Interface.php#L76-L88